tools/inference_engine/src/impl/model.cpp

```diff
@@ -178,6 +184,46 @@ ModelDescriptor::~ModelDescriptor()
     std::cout << "D-TOR ~ModelDescriptor()" << std::endl;
 }
 
+class FusionVisitor : public GpuVisitor {
+
+    std::unordered_set<GpuNode*> to_delete_;
+public:
+    void processSortedNodes(std::vector<std::unique_ptr<GpuNode>>& sortedNodes) override {
+        to_delete_.clear();
+
+        for (auto&& node : sortedNodes) {
+            node->accept(this);
+        }
+
+        if (!to_delete_.empty()) {
+            std::cout << "FusionVisitor: removing nodes...\n";
+
+            std::erase_if(sortedNodes, [&](auto& el) {return to_delete_.contains(el.get()); });
+        }
+    }
+
+    virtual void visit(GpuPort* pn) override {
+        std::cout << "visiting port...\n";
+    }
+    virtual void visit(GpuActivation* pn) override {
+        std::cout << "visiting activation...\n";
+
+        // check matmul + activation and fuse?
+        auto& inputs = pn->get_inputs();
+        if (inputs.size() == 1) {
+            auto possibleMatMul = dynamic_cast<GpuMatMul*>(inputs[0]);
+            if (possibleMatMul && possibleMatMul->get_outputs().size() == 1) {
+                std::cout << "possible matmul + activation fusion...\n";
+                possibleMatMul->fuse_with(pn);
+                to_delete_.insert(pn);
```

Comment:
Shouldn't `pn` be deleted immediately? What if we have chain of layers which can be fused. For example acitvation -> activation -> activation -> etc...?